packages/vscode-messenger-common/src/messages.ts

```diff
@@ -176,8 +176,83 @@ export type NotificationHandler<P> = (params: P, sender: MessageParticipant) =>
  * Base API for Messenger implementations.
  */
 export interface MessengerAPI {
-    sendRequest<P, R>(type: RequestType<P, R>, receiver: MessageParticipant, params?: P): Promise<R>
+    sendRequest<P, R>(type: RequestType<P, R>, receiver: MessageParticipant, params?: P, cancelable?: Cancelable): Promise<R>
     onRequest<P, R>(type: RequestType<P, R>, handler: RequestHandler<P, R>): void
     sendNotification<P>(type: NotificationType<P>, receiver: MessageParticipant, params?: P): void
     onNotification<P>(type: NotificationType<P>, handler: NotificationHandler<P>): void
 }
+
+/**
+ *  Deferred promise that can be resolved or rejected later.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export class PendingRequest<R = any> {
+    resolve: (value: R) => void;
+    reject: (reason?: unknown) => void;
+
+    result = new Promise<R>((resolve, reject) => {
+        this.resolve = (arg) => resolve(arg);
+        this.reject = (err) => reject(err);
+    });
+}
+
+/**
+ * Interface that allows to check for cancellation and
+ * set a listener that is called when the request is canceled.
+ */
+export interface CancelIndicator {
+    isCanceled(): boolean;
+    onCancel: ((reason: string) => void) | undefined;
```

Comment:
This should be a function accepting a callback.

Reply:
Mostly in JS **on**_Xyz_ are properties to set. Functions that accepts a callback are called **add**_XyzListener_.
In my first version I had a `addCancelListener((reason: string) => void)`  but I liked the current approach better. What is the benefit of `onCancel((reason: string) => void)` compared to the current implementation?

Reply:
> What is the benefit of `onCancel((reason: string) => void)` compared to the current implementation?

You can have multiple listeners, not just a single one.

Reply:
Handlers are normally not shared among different components so there is no need to add multiple cancel listeners. If there should be such a request, we can still change that by turning `onCancel ` to a setter:
```ts

private listeners =  new Set<((reason: string) => void)>();

set onCancel(listener: (reason: string) => void) {
    this.listeners.add(listener);
}

public cancel(reason: string): void {
    if (this.canceled) {
        throw new Error('Request already canceled');
    }
    this.canceled = true;
    this.listeners.forEach(l => l(reason));
}
``` 

WDYT?

Reply:
A request handler can trigger a complex task involving multiple asynchronous steps. The cancellation token is passed to each of these steps so they can react to cancellation events. So there is definitely the need to register multiple cancellation listeners.

Pseudo code:
```typescript
messenger.onRequest(type, async (params, token) => {
    const data1 = await service1(params, token); // service1 listens to cancellation events while fetching data
    const data2 = await service2(params, token); // service2 does the same
    return processData(data2, data2);
});
```

Reply:
`onCancel: (reason: string) => void` field is now replaced with `addCancelListener((reason: string) => void)` function